cmd/thanos/rule.go

```diff
@@ -200,6 +201,55 @@ func registerRule(m map[string]setupFunc, app *kingpin.Application) {
 	}
 }
 
+// RuleMetrics defines thanos rule metrics.
+type RuleMetrics struct {
+	configSuccess     prometheus.Gauge
+	configSuccessTime prometheus.Gauge
+	duplicatedQuery   prometheus.Counter
+	rulesLoaded       *prometheus.GaugeVec
+	ruleEvalWarnings  *prometheus.CounterVec
+}
+
+func newRuleMetrics(reg *prometheus.Registry) *RuleMetrics {
+	m := new(RuleMetrics)
+
+	m.configSuccess = prometheus.NewGauge(prometheus.GaugeOpts{
+		Name: "thanos_rule_config_last_reload_successful",
+		Help: "Whether the last configuration reload attempt was successful.",
+	})
+	m.configSuccessTime = prometheus.NewGauge(prometheus.GaugeOpts{
+		Name: "thanos_rule_config_last_reload_success_timestamp_seconds",
+		Help: "Timestamp of the last successful configuration reload.",
+	})
+	m.duplicatedQuery = prometheus.NewCounter(prometheus.CounterOpts{
+		Name: "thanos_rule_duplicated_query_addresses_total",
+		Help: "The number of times a duplicated query addresses is detected from the different configs in rule",
+	})
+	m.rulesLoaded = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Name: "thanos_rule_loaded_rules",
+			Help: "Loaded rules partitioned by file and group",
+		},
+		[]string{"strategy", "file", "group"},
+	)
+	m.ruleEvalWarnings = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "thanos_rule_evaluation_with_warnings_total",
+			Help: "The total number of rule evaluation that were successful but had warnings which can indicate partial error.",
+		}, []string{"strategy"},
+	)
+	m.ruleEvalWarnings.WithLabelValues(strings.ToLower(storepb.PartialResponseStrategy_ABORT.String()))
+	m.ruleEvalWarnings.WithLabelValues(strings.ToLower(storepb.PartialResponseStrategy_WARN.String()))
+
+	reg.MustRegister(m.configSuccess)
+	reg.MustRegister(m.configSuccessTime)
+	reg.MustRegister(m.duplicatedQuery)
+	reg.MustRegister(m.rulesLoaded)
+	reg.MustRegister(m.ruleEvalWarnings)
```

Comment:
Better to check if the registry is nil or not.
```suggestion
if (reg != nil) {
	reg.MustRegister(m.configSuccess)
	reg.MustRegister(m.configSuccessTime)
	reg.MustRegister(m.duplicatedQuery)
	reg.MustRegister(m.rulesLoaded)
	reg.MustRegister(m.ruleEvalWarnings)
}
```